tests/utils/test_shared_process_pool.py

```diff
@@ -0,0 +1,291 @@
+    with pytest.raises(TypeError):
+        pool.submit_task("test_stage", _arbitrary_function, threading.Lock())
+
+
+@pytest.mark.parametrize(
+    "a, b, expected",
+    [
+        (1, 2, 3),  # Integers
+        (complex(1, 2), complex(3, 4), complex(4, 6)),  # Complex numbers
+        (Decimal('1.1'), Decimal('2.2'), Decimal('3.3')),  # Decimal numbers
+        (Fraction(1, 2), Fraction(1, 3), Fraction(5, 6)),  # Fractions
+        ("Hello, ", "world!", "Hello, world!"),  # Strings
+        ([1, 2, 3], [4, 5, 6], [1, 2, 3, 4, 5, 6]),  # Lists
+        ((1, 2, 3), (4, 5, 6), (1, 2, 3, 4, 5, 6)),  # Tuples
+    ])
+def test_submit_multiple_tasks(shared_process_pool, a, b, expected):
+
+    pool = shared_process_pool
+    pool.set_usage("test_stage", 0.5)
+
+    num_tasks = 100
+    tasks = []
+    for _ in range(num_tasks):
+        tasks.append(pool.submit_task("test_stage", _add_task, a, b))
+
+    for future in tasks:
+        assert future.result() == expected
+
+
+def test_set_usage(shared_process_pool):
+
+    pool = shared_process_pool
+    pool.wait_until_ready()
+
+    pool.set_usage("test_stage_1", 0.5)
+    assert pool._total_usage == 0.5
+    _check_pool_stage_settings(pool, "test_stage_1", 0.5)
+
+    pool.set_usage("test_stage_2", 0.3)
+    assert pool._total_usage == 0.8
+    _check_pool_stage_settings(pool, "test_stage_2", 0.3)
+
+    # valid update to the usage of an existing stage
+    pool.set_usage("test_stage_1", 0.6)
+    assert pool._total_usage == 0.9
+    _check_pool_stage_settings(pool, "test_stage_1", 0.6)
+
+    # invalid update to the usage of an existing stage, exceeding the total usage limit
+    with pytest.raises(ValueError):
+        pool.set_usage("test_stage_1", 0.8)
+
+    # adding a new stage usage, exceeding the total usage limit
+    with pytest.raises(ValueError):
+        pool.set_usage("test_stage_3", 0.2)
+
+    with pytest.raises(ValueError):
+        pool.set_usage("test_stage_1", 1.1)
+
+    with pytest.raises(ValueError):
+        pool.set_usage("test_stage_1", -0.1)
+
+    # invalid settings should not change the pool status
+    _check_pool_stage_settings(pool, "test_stage_1", 0.6)
+    assert pool._total_usage == 0.9
+
+
+def test_task_completion_with_early_stop(shared_process_pool):
+
+    pool = shared_process_pool
+    pool.set_usage("test_stage_1", 0.1)
+    pool.set_usage("test_stage_2", 0.3)
+    pool.set_usage("test_stage_3", 0.5)
+
+    tasks = []
+
+    task_num = 5
+    sleep_time = 10
+    for _ in range(task_num):
+        tasks.append(pool.submit_task("test_stage_1", _time_consuming_task, sleep_time))
+        tasks.append(pool.submit_task("test_stage_2", _time_consuming_task, sleep_time))
+        tasks.append(pool.submit_task("test_stage_3", _time_consuming_task, sleep_time))
+
+    pool.stop()
+    pool.join()
+
+    # all tasks should be completed before the pool is shutdown
+    assert len(tasks) == 3 * task_num
+    for task in tasks:
+        assert task.done()
```

Comment:
Instead of using sleep which is terrible for testing, you can exercise the same functionality like this:

```suggestion
def block_until_signaled(q: multiprocessing.Queue):
    return q.get()


def test_task_completion_with_early_stop(shared_process_pool: SharedProcessPool):

    pool = shared_process_pool
    pool.set_usage("test_stage_1", 0.1)
    pool.set_usage("test_stage_2", 0.3)
    pool.set_usage("test_stage_3", 0.5)

    m = multiprocessing.Manager()
    queue = m.Queue()

    tasks = []

    task_num = 5
    sleep_time = 10
    for _ in range(task_num):
        tasks.append(pool.submit_task("test_stage_1", block_until_signaled, queue))
        tasks.append(pool.submit_task("test_stage_2", block_until_signaled, queue))
        tasks.append(pool.submit_task("test_stage_3", block_until_signaled, queue))

    pool.stop()

    # Ensure that all tasks have not completed since they havent been signaled yet
    for task in tasks:
        assert not task.done()

    # Signal the tasks to complete
    for i in range(len(tasks)):
        queue.put(i)

    pool.join()

    # all tasks should be completed before the pool is shutdown
    assert len(tasks) == 3 * task_num
    for task in tasks:
        assert task.done()
```

Note: This does require some changes to how `stop` and `join` function. See comments on those functions